File: source/annotate-numbers.py
#!/usr/bin/env python
#! nix-shell -i python -p "python38.withPackages(p:[p.pillow])"

import sys
from PIL import Image


class Img:
    def __init__(self, fname, zoom):
        self._img = Image.open(fname)
        self._pixels = self._img.load()
        self._zoom = zoom

        self.size = self._img.size[0] // zoom, self._img.size[1] // zoom

    def __getitem__(self, xy):
        xy = xy[0] * self._zoom, xy[1] * self._zoom
        try:
            c = self._pixels[xy]
        except IndexError:
            return False
        return c[0] + c[1] + c[2] > 382


class Svg:
    def __init__(self, fname, width, height):
        self._f = open(fname, "w")
        self._print(
            f'<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="{width*8}" height="{height*8}">'
        )
        self._print(f'<rect width="{width*8}" height="{height*8}" style="fill:black"/>')

    def _print(self, *args, **kwargs):
        print(*args, **kwargs, file=self._f)

    def point(self, x, y):
        self._print(
            f'<rect x="{x*8}" y="{y*8}" width="7" height="7" style="fill:white"/>'
        )

    def annotation(self, x, y, w, h, text):
        self._print(
            f'<rect x="{x*8}" y="{y*8}" width="{w*8}" height="{h*8}" style="fill:green;opacity:0.5"/>'
        )
        style = "paint-order: stroke; fill: white; stroke: black; stroke-width: 2px; font:24px bold sans;"
        self._print(
            f'<text x="{x*8+w*4}" y="{y*8+h*4}" dominant-baseline="middle" text-anchor="middle" fill="white" style="{style}">{text}</text>'
        )

    def close(self):
        self._print("</svg>")
        self._f.close()


def decode_number(img, x, y):
    if img[x - 1, y - 1] or img[x, y - 1] or img[x - 1, y] or img[x, y]:
        return None

    # Get the size by iterating over top and left edges
    size = 0
    while True:
        items = (
            img[x + size + 1, y - 1],
            img[x + size + 1, y],
            img[x - 1, y + size + 1],
            img[x, y + size + 1],
        )
        if items == (False, True, False, True):
            size += 1
            continue
        if items == (False, False, False, False):
            break
        return None

    if size == 0:
        return None

    # Check that right and bottom edges are empty
    for i in range(size + 2):
        if img[x + size + 1, y] or img[x, y + size + 1]:
            return None

    # Decode the number
    result, d = 0, 1
    for iy in range(size):
        for ix in range(size):
            result += d * img[x + ix + 1, y + iy + 1]
            d *= 2

    return size, result


def main(in_fname, out_fname):
    img = Img(in_fname, 4)
    svg = Svg(out_fname, img.size[0], img.size[1])

    for y in range(img.size[1]):
        for x in range(img.size[0]):
            if img[x, y]:
                svg.point(x, y)

    for y in range(img.size[1]):
        for x in range(img.size[0]):
            if (n := decode_number(img, x, y)) is not None:
                svg.annotation(x - 0.5, y - 0.5, n[0] + 2, n[0] + 2, n[1])
    svg.close()


main(sys.argv[1], sys.argv[2])

File: source/annotate-numbers.hs
#! /usr/bin/env nix-shell
#! nix-shell -i runhaskell -p
#! nix-shell "haskellPackages.ghcWithPackages (pkgs: with pkgs; [JuicyPixels JuicyPixels-util errors])"

import qualified Codec.Picture as P
import qualified Codec.Picture.RGBA8 as P8
import Data.Word (Word16)
import Data.List
import Debug.Trace
import Control.Error.Safe
import Safe (lastMay)
import Data.Maybe (catMaybes)
import System.Environment (getArgs)

--------------------------------------------------------------------------------
-- Img

type Scale = Int
type Coord = (Int, Int)
type Size = (Int, Int)
data Img = Img (P.Image P.PixelRGBA8) Scale

imgLoad :: FilePath -> Scale -> IO Img
imgLoad path scale = (\img -> Img img scale) <$> P8.readImageRGBA8 path

imgWidth :: Img -> Int
imgWidth (Img img scale) = P.imageWidth img `div` scale

imgHeight :: Img -> Int
imgHeight (Img img scale) = P.imageHeight img `div` scale

imgPixel :: Img -> Coord -> Bool
imgPixel (Img img scale) (x, y) =
  if x' < 0 || y' < 0 || x' >= P.imageWidth img || y' >= P.imageHeight img
  then False
  else fromIntegral r + fromIntegral g + fromIntegral b > (0::Word16)
  where
    x' = x * scale
    y' = y * scale
    P.PixelRGBA8 r g b _ = P.pixelAt img x' y'

imgDump :: Img -> [Int] -> [Int] -> String
imgDump img xs ys = unlines $ map showLine ys
  where showLine y = concatMap (\x -> if imgPixel img (x,y) then "#" else ".") xs

imgDumpFull :: Img -> String
imgDumpFull img = imgDump img [0..imgWidth img - 1] [0..imgHeight img - 1]

imgAllPixels :: Img -> [Coord]
imgAllPixels img = [(x, y) | x <- [0..imgWidth img - 1], y <- [0..imgHeight img - 1]]

--------------------------------------------------------------------------------
-- Number decoder

decodeNumber :: Img -> Coord -> Maybe (Integer, Size)
decodeNumber img (x, y) = do
  {-
    Figure:
       . . _ _ _ _ ,
       . : # # # # ,   → x
       _ # + + + + -
       _ # + + + + -
       _ # + + + + -
       _ # + + + + -
       , ? - - - - -

         ↓
         y

     Legend:
       . : _ , - — black pixels
       #         — white pixels
       ?         — negativity bit
       +         — binary data
       :         — point (x, y)
  -}

  let px = imgPixel img

  -- 1. Check that 2x2 top left corner is empty (`.` and `:`)
  assertMay $ not $ any px [(x-1, y-1), (x, y-1), (x-1, y), (x,y)]

  -- 2. Calculate the size based on top and left edges (`.` and `#`)
  let topLeft' i = (px (x + i, y - 1), px (x + i, y),
                    px (x - 1, y + i), px (x,     y + i))
  let topLeft = takeWhile (\i -> (False, True, False, True) == topLeft' i) $ [1..]
  size <- lastMay topLeft
  assertMay $ size >= 1

  -- 3. Check the negativity bit (`?`) and empty space at corners (`,`)
  negative <-
    case topLeft' (size+1) of
      (False, False, False, False) -> Just False
      (False, False, False, True) -> Just True
      otherwise -> Nothing

  -- 4. Check that right and bottom edges are empty (`-`)
  assertMay $ not $
    any (\i -> px (x + size + 1, y+i) || px(x+i, y + size + 1)) [1 .. size+1]

  -- 5. Decode binary data
  let number = bitsToInteger [px (ix,iy) | iy <- [y+1 .. y+size], ix <- [x+1 .. x+size]]

  Just $ if negative
    then (-number, (size, size+1))
    else (number, (size, size))

bitsToInteger :: [Bool] -> Integer
bitsToInteger = fst . foldl' f (0, 1)
  where
    f (sum, bit) True = (sum + bit, bit*2)
    f (sum, bit) False = (sum, bit*2)

--------------------------------------------------------------------------------
-- svg

svg img annotations =
  concat $ (
    svgHead img ++
    svgImgPoints img ++
    concatMap (\(x,y,w,h,text) -> svgAnnotation x y w h text) annotations ++
    ["</svg>"]
  )

svgHead img = [
    "<svg xmlns='http://www.w3.org/2000/svg' version='1.1' width='",
    show $ imgWidth img * 8,
    "' height='",
    show $ imgHeight img * 8,
    "'>\n",
    "<rect width='100%' height='100%' style='fill:black'/>\n"
  ]

svgPoint img (x, y) = [
    "<rect x='", show (x*8),
    "' y='", show (y*8),
    "' width='7' height='7' style='fill:white'/>\n"
  ]

svgImgPoints img = 
  concatMap (svgPoint img) $ filter (imgPixel img) $ imgAllPixels img

svgAnnotation :: Double -> Double -> Double -> Double -> String -> [String]
svgAnnotation x y w h text = [
    "<rect x='", show (x*8),
    "' y='", show (y*8),
    "' width='", show (w*8),
    "' height='", show (h*8),
    "' style='fill:green;opacity:0.5'/>\n",

    "<text x='", show (x*8 + w*4),
    "' y='", show (y*8 + h*4),
    "' dominant-baseline='middle' text-anchor='middle' fill='white' style='",
    "paint-order: stroke; fill: white; stroke: black; stroke-width: 2px; font:24px bold sans;",
    "'>", text, "</text>\n"
  ]


--------------------------------------------------------------------------------
-- Main

decodeAnnotate img (x,y) = do
  (number, (w, h)) <- decodeNumber img (x, y)
  Just (fromIntegral x - 0.5, fromIntegral y - 0.5, fromIntegral w+2, fromIntegral h+2, show number)

annotateImg :: Img -> String
annotateImg img = svg img annotations
  where
    annotations = catMaybes $ map (decodeAnnotate img) $ imgAllPixels img

main = do
  [fnameIn, fnameOut] <- getArgs
  img <- imgLoad fnameIn 4
  writeFile fnameOut $ annotateImg img
